File: backend/src/main/java/com/hostel/backend/service/AdminService.java
package com.hostel.backend.service;

import com.hostel.backend.dto.*;
import com.hostel.backend.entity.Department;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.ComplaintStatus;
import com.hostel.backend.enums.Role;
import com.hostel.backend.repository.ComplaintRepository;
import com.hostel.backend.repository.DepartmentRepository;
import com.hostel.backend.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class AdminService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final ComplaintRepository complaintRepository;
    private final DepartmentRepository departmentRepository; // ✅ NEW

    public AdminService(UserRepository userRepository,
                        PasswordEncoder passwordEncoder,
                        ComplaintRepository complaintRepository,
                        DepartmentRepository departmentRepository) { // ✅ NEW
        this.userRepository       = userRepository;
        this.passwordEncoder      = passwordEncoder;
        this.complaintRepository  = complaintRepository;
        this.departmentRepository = departmentRepository; // ✅ NEW
    }

    // ── User Creation ─────────────────────────────────────────────

    public User createUserWithRole(RegisterRequest request, Role role) {
        if (request.getName() == null || request.getEmail() == null ||
            request.getPassword() == null || request.getPhoneNumber() == null) {
            throw new RuntimeException("Required fields are missing");
        }
        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new RuntimeException("User with this email already exists");
        }

        User.UserBuilder builder = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(role)
                .phoneNumber(request.getPhoneNumber())
                .active(true);

        // ✅ NEW — assign department if provided (for DEPT_HEAD)
        if (request.getDepartmentId() != null) {
            Department dept = departmentRepository.findById(request.getDepartmentId())
                    .orElseThrow(() -> new RuntimeException("Department not found"));
            builder.department(dept);
        }

        return userRepository.save(builder.build());
    }

    // ── User Queries ──────────────────────────────────────────────

    public List<AdminUserResponse> getAllUsers() {
        return userRepository.findAll()
                .stream()
                .map(this::mapToAdminUserResponse)
                .collect(Collectors.toList());
    }

    public List<AdminUserResponse> getUsersByRole(Role role) {
        return userRepository.findAll()
                .stream()
                .filter(u -> u.getRole() == role)
                .map(this::mapToAdminUserResponse)
                .collect(Collectors.toList());
    }

    // ── User Updates ──────────────────────────────────────────────

    public User updateUser(Long id, UpdateUserRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        if (request.getName() != null)        user.setName(request.getName());
        if (request.getPhoneNumber() != null)  user.setPhoneNumber(request.getPhoneNumber());
        if (request.getRoomNumber() != null)   user.setRoomNumber(request.getRoomNumber());
        return userRepository.save(user);
    }

    public void deactivateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        user.setActive(false);
        userRepository.save(user);
    }

    public void activateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        user.setActive(true);
        userRepository.save(user);
    }

    public void deleteUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        userRepository.delete(user);
    }

    // ── Departments ───────────────────────────────────────────────

    // ✅ NEW
    public List<Department> getAllDepartments() {
        return departmentRepository.findAll();
    }

    // ── Stats ─────────────────────────────────────────────────────

    public DashboardStatsResponse getDashboardStats() {
        long total      = complaintRepository.count();
        long created    = complaintRepository.countByStatus(ComplaintStatus.CREATED);
        long assigned   = complaintRepository.countByStatus(ComplaintStatus.ASSIGNED);
        long inProgress = complaintRepository.countByStatus(ComplaintStatus.IN_PROGRESS);
        long resolved   = complaintRepository.countByStatus(ComplaintStatus.RESOLVED);
        long closed     = complaintRepository.countByStatus(ComplaintStatus.CLOSED);
        long rejected   = complaintRepository.countByStatus(ComplaintStatus.REJECTED);
        return new DashboardStatsResponse(total, created, assigned, inProgress,
                                          resolved, closed, rejected);
    }

    private AdminUserResponse mapToAdminUserResponse(User user) {
        Double averageRating = null;
        if (user.getRole() == Role.WORKER) {
            Double raw = complaintRepository.findAverageRatingByWorker(user);
            if (raw != null) averageRating = Math.round(raw * 10.0) / 10.0;
        }
        return AdminUserResponse.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .phoneNumber(user.getPhoneNumber())
                .scholarNumber(user.getScholarNumber())
                .hostelBlock(user.getBlock() != null ? user.getBlock().getName() : null)
                .roomNumber(user.getRoomNumber())
                .department(user.getDepartment() != null ? user.getDepartment().getName() : null)
                .active(user.isActive())
                .averageRating(averageRating)
                .build();
    }
}

File: backend/src/services/admin.service.js
const prisma = require('../config/prisma');
const { AppError, NotFoundError } = require('../utils/AppError');
const { hashPassword } = require('../utils/password');
const { omitPassword } = require('../utils/mappers');
const authService = require('./auth.service');

// Admin-created student accounts skip the OTP/self-verification flow and
// are active immediately — same as admin-created warden/caretaker/worker
// accounts always have been. Distinct from the public self-registration
// endpoint (`/auth/register`), which still requires OTP verification.
async function createStudent(request) {
  const user = await authService.createStudentAccount(request, { sendOtp: false });
  return omitPassword(user);
}

// Mirrors AdminService.java `createUserWithRole()`.
async function createUserWithRole(request, role) {
  const { name, email, password, phoneNumber, departmentId } = request;
  if (!name || !email || !password || !phoneNumber) {
    throw new AppError('Required fields are missing', 400);
  }

  const existing = await prisma.user.findUnique({ where: { email } });
  if (existing) throw new AppError('User with this email already exists', 400);

  let department;
  if (departmentId) {
    department = await prisma.department.findUnique({ where: { id: BigInt(departmentId) } });
    if (!department) throw new AppError('Department not found', 400);
  }

  const user = await prisma.user.create({
    data: {
      name,
      email,
      password: await hashPassword(password),
      role,
      phoneNumber,
      active: true,
      appEnabled: true,
      departmentId: department ? department.id : undefined,
    },
  });
  return omitPassword(user);
}

async function averageRatingByWorker(workerId) {
  const result = await prisma.complaint.aggregate({
    where: { workerId, rating: { not: null } },
    _avg: { rating: true },
  });
  return result._avg.rating;
}

async function mapToAdminUserResponse(user) {
  let averageRating = null;
  if (user.role === 'WORKER') {
    const raw = await averageRatingByWorker(user.id);
    if (raw != null) averageRating = Math.round(raw * 10) / 10;
  }
  return {
    id: user.id,
    name: user.name,
    email: user.email,
    role: user.role,
    phoneNumber: user.phoneNumber,
    scholarNumber: user.scholarNumber,
    hostelBlock: user.block || null,
    roomNumber: user.roomNumber,
    department: user.department ? user.department.name : null,
    active: user.active,
    averageRating,
  };
}

async function getAllUsers() {
  const users = await prisma.user.findMany({ include: { department: true } });
  return Promise.all(users.map(mapToAdminUserResponse));
}

async function getUsersByRole(role) {
  const users = await prisma.user.findMany({
    where: { role },
    include: { department: true },
  });
  return Promise.all(users.map(mapToAdminUserResponse));
}

async function updateUser(id, request) {
  const user = await prisma.user.findUnique({ where: { id: BigInt(id) } });
  if (!user) throw new NotFoundError('User not found');

  const data = {};
  if (request.name != null) data.name = request.name;
  if (request.phoneNumber != null) data.phoneNumber = request.phoneNumber;
  if (request.roomNumber != null) data.roomNumber = request.roomNumber;

  const saved = await prisma.user.update({ where: { id: user.id }, data });
  return omitPassword(saved);
}

async function deactivateUser(id) {
  const user = await prisma.user.findUnique({ where: { id: BigInt(id) } });
  if (!user) throw new AppError(`User not found with id: ${id}`, 400);
  await prisma.user.update({ where: { id: user.id }, data: { active: false } });
}

async function activateUser(id) {
  const user = await prisma.user.findUnique({ where: { id: BigInt(id) } });
  if (!user) throw new AppError(`User not found with id: ${id}`, 400);
  await prisma.user.update({ where: { id: user.id }, data: { active: true } });
}

async function deleteUser(id) {
  const user = await prisma.user.findUnique({ where: { id: BigInt(id) } });
  if (!user) throw new AppError('User not found', 400);
  await prisma.user.delete({ where: { id: user.id } });
}

async function getAllDepartments() {
  return prisma.department.findMany();
}

// Mirrors AdminService.java `getDashboardStats()` EXACTLY — including its
// pre-existing field-order bug (this is a *different* method from
// ComplaintService.getDashboardStats(), used by GET /api/admin/dashboard-stats
// rather than GET /api/complaints/dashboard). The original passes
// (total, created, assigned, inProgress, resolved, closed, rejected, 0) into
// a DTO whose 6th/7th/8th fields are actually (disputed, closed, rejected),
// so `disputed` reports the closed-count, `closed` reports the
// rejected-count, and `rejected` is always 0. See
// docs/MIGRATION_NOTES.md "Known gaps carried over" — preserved verbatim
// because this task's constraint is 100% behavioral parity, not bug fixes.
async function getDashboardStats() {
  const [total, created, assigned, inProgress, resolved, closed, rejected] = await Promise.all([
    prisma.complaint.count(),
    prisma.complaint.count({ where: { status: 'CREATED' } }),
    prisma.complaint.count({ where: { status: 'ASSIGNED' } }),
    prisma.complaint.count({ where: { status: 'IN_PROGRESS' } }),
    prisma.complaint.count({ where: { status: 'RESOLVED' } }),
    prisma.complaint.count({ where: { status: 'CLOSED' } }),
    prisma.complaint.count({ where: { status: 'REJECTED' } }),
  ]);
  return {
    total,
    created,
    assigned,
    inProgress,
    resolved,
    disputed: closed, // scrambled — see note above
    closed: rejected, // scrambled — see note above
    rejected: 0, // scrambled — see note above
  };
}

module.exports = {
  createStudent,
  createUserWithRole,
  getAllUsers,
  getUsersByRole,
  updateUser,
  deactivateUser,
  activateUser,
  deleteUser,
  getAllDepartments,
  getDashboardStats,
  averageRatingByWorker,
};
